plugins/intrinsics/src/main/java/org/qbicc/plugin/intrinsics/core/CoreIntrinsics.java

```diff
@@ -948,9 +952,17 @@ static void registerOrgQbiccRuntimeCNativeIntrinsics(final CompilationContext ct
 
         InstanceIntrinsic identity = (builder, instance, target, arguments) -> instance;
 
-        intrinsics.registerIntrinsic(ptrDesc, "deref", MethodDescriptor.synthesize(classContext, nObjDesc, List.of()), identity);
         intrinsics.registerIntrinsic(ptrDesc, "asArray", MethodDescriptor.synthesize(classContext, ArrayTypeDescriptor.of(classContext, nObjDesc), List.of()), identity);
 
+        InstanceIntrinsic deref = (builder, instance, target, arguments) -> {
+            MemoryAtomicityMode mode = MemoryAtomicityMode.UNORDERED;
+            if (instance.getType() instanceof PointerType pt && pt.getPointeeType() instanceof CompoundType) {
+                mode = MemoryAtomicityMode.NONE;
+            }
+            return builder.deref(builder.load(builder.pointerHandle(instance), mode));
```

Comment:
I'm still not sure where this load comes into play as it would be needed for any case that is not a compound type accessing a field.

Reply:
Honestly I'm no longer sure that it is even possible to do this correctly in this way.  The problem is that loads must always be ordered, so they cannot migrate out from their original position (at least, until we have better dependency links between memory operations). At the same time, we want `Deref` to be side-effect free i.e. not ordered. These two requirements seem fundamentally incompatible; either `Deref` becomes a redundant variation of `Load` or else we have invalid ordering. 
Even if we generate a load to keep in our back pocket in the event that we need it to be transformed, then the load is in the graph already and the `Deref` is not really doing anything useful.

We might have to pivot to a completely different approach, like abandoning `Deref` and transforming either or both of:

* `Load(MemberOf(Load(MemberOf(x, m1), NONE), m2), NONE)` -> `Load(MemberOf(MemberOf(x, m1), m2), NONE)`
* `ExtractMember(Load(x, NONE), m1)` -> `Load(MemberOf(x, m1), NONE)` (this one is slightly complicated by the fact that the new load has to be placed with the original load due to ordering, which requires coordination between two phases: one to discover the member extraction and one to generate the load)

...and then sweeping up the unused whole-structure loads using a new load elimination pass.  (Also apply similar transformations to array value nodes.)

Reply:
I think you're right. I've updated this pr then to remove deref and put back `load(pointervalue(instance))` for the deref intrinsic. Combined with your recent MethodParser changes to use ExtractMember this is working fine to access compound members.
Since the transformations you mentioned would be optimizations I can update the orignal issues for those in a separate pull request.

Reply:
Sounds good. We can revisit later if we need to optimize it more for any reason.